File: ruby/lib/solana_mpp/server/verification_result.rb
# frozen_string_literal: true

module SolanaMpp
  module Server
    # Result returned by lower-level credential verifiers.
    class VerificationResult
      attr_reader :reference, :reason, :credential, :challenge

      def initialize(ok:, reference: nil, reason: nil, credential: nil, challenge: nil)
        @ok = ok
        @reference = reference
        @reason = reason
        @credential = credential
        @challenge = challenge
      end

      # Return true when verification succeeded.
      def ok?
        @ok
      end

      # Create a successful verification result.
      def self.success(reference: nil, credential: nil, challenge: nil)
        new(ok: true, reference: reference, credential: credential, challenge: challenge)
      end

      # Create a failed verification result.
      def self.failure(reason)
        new(ok: false, reason: reason)
      end
    end
  end
end

File: php/src/Protocols/Mpp/Server/VerificationResult.php
<?php

declare(strict_types=1);

namespace PayKit\Protocols\Mpp\Server;

use PayKit\Protocols\Mpp\Core\Challenge;
use PayKit\Protocols\Mpp\Core\Credential;

/**
 * Carries the result of a payment credential verification attempt.
 *
 * On success, when the result was produced by {@see ChargeServer::verifyAuthorizationHeader()},
 * `challenge` and `credential` are populated with the verified payloads so
 * callers can mint a receipt without re-parsing the Authorization header.
 */
final class VerificationResult
{
    /**
     * Create an immutable verification result.
     */
    private function __construct(
        public readonly bool $ok,
        public readonly string $reason = '',
        public readonly string $reference = '',
        public readonly string $externalId = '',
        public readonly ?Challenge $challenge = null,
        public readonly ?Credential $credential = null,
    ) {
    }

    /**
     * Return a successful verification result with its settlement reference.
     */
    public static function success(string $reference, string $externalId = ''): self
    {
        return new self(ok: true, reference: $reference, externalId: $externalId);
    }

    /**
     * Return a failed verification result with a developer-readable reason.
     */
    public static function failure(string $reason): self
    {
        return new self(ok: false, reason: $reason);
    }

    /**
     * Return a copy with the verified challenge and credential attached.
     *
     * Intended for internal use by {@see ChargeServer} after a verifier
     * accepts a credential, so the receipt step does not need to re-parse the
     * Authorization header.
     */
    public function withVerified(Challenge $challenge, Credential $credential): self
    {
        return new self(
            ok: $this->ok,
            reason: $this->reason,
            reference: $this->reference,
            externalId: $this->externalId,
            challenge: $challenge,
            credential: $credential,
        );
    }
}
